File: frontend/src/types/asset.ts
export interface LatestPrice {
  close: number;
  change_pct: number | null;
  date: string;
}

export interface AssetResponse {
  id: number;
  symbol: string;
  name_ja: string | null;
  asset_type: string;
  market: string;
  is_active: boolean;
  latest_price: LatestPrice | null;
}

export interface AssetPricePoint {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  adj_close: number | null;
  volume: number | null;
}

export interface AssetPricesResponse {
  symbol: string;
  prices: AssetPricePoint[];
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

File: backend/src/app/schemas/asset.py
"""Asset-related Pydantic schemas."""

from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class LatestPrice(BaseModel):
    close: float
    date: date
    change_pct: float | None = None


class AssetResponse(BaseModel):
    symbol: str
    name: str
    name_ja: str | None = None
    asset_type: str
    market: str
    currency: str
    sector: str | None = None
    latest_price: LatestPrice | None = None

    model_config = {"from_attributes": True}


class AssetPriceResponse(BaseModel):
    date: date
    open: float | None = None
    high: float | None = None
    low: float | None = None
    close: float
    adj_close: float | None = None
    volume: int | None = None

    model_config = {"from_attributes": True}


class AssetPricesResponse(BaseModel):
    symbol: str
    period: str
    interval: str
    prices: list[AssetPriceResponse]
    data_source_note: str = "データは教育・情報提供目的です。投資判断の根拠として使用しないでください。"


class PaginatedAssetResponse(BaseModel):
    items: list[AssetResponse]
    total: int
    page: int
    per_page: int
    pages: int
